packages/contracts/src/token/ERC1155/RedeemOptimizerFIFOLIFO.sol

```diff
@@ -0,0 +1,299 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.20;
+
+import { IMultiTokenVault } from "@credbull/token/ERC1155/IMultiTokenVault.sol";
+import { IRedeemOptimizer } from "@credbull/token/ERC1155/IRedeemOptimizer.sol";
+import { Math } from "@openzeppelin/contracts/utils/math/Math.sol";
+
+import { console } from "forge-std/console.sol";
+
+/**
+ * @title The redemption optimizer utilising a combined FIFO/LIFO strategy.
+ * @notice Optimizes the redemption of shares using a FIFO strategy with a LIFO component.
+ * @dev The strategy is applied FIFO first and, if needed, LIFO. FIFO selects mature deposits to redeem to maximise
+ *  the value returned while minimising deposits redeemed. If more value is required, then LIFO selects the most recent
+ *  first deposits, to make up the value and reduce the loss of invested time in the older deposits.
+ */
+contract RedeemOptimizerFIFOLIFO is IRedeemOptimizer {
+    using Math for uint256;
+
+    error RedeemOptimizer__InvalidDepositPeriodRange(uint256 fromPeriod, uint256 toPeriod);
+    error RedeemOptimizer__FutureToDepositPeriod(uint256 toPeriod, uint256 currentPeriod);
+    error RedeemOptimizer__OptimizerFailed(uint256 amountFound, uint256 amountToFind);
+
+    OptimizerBasis public immutable DEFAULT_BASIS;
+    uint256 public immutable START_DEPOSIT_PERIOD;
+    uint256 public immutable TENOR;
+
+    uint256 private _startDepositPeriod;
+
+    constructor(OptimizerBasis defaultBasis, uint256 startDepositPeriod, uint256 tenor) {
+        DEFAULT_BASIS = defaultBasis;
+        START_DEPOSIT_PERIOD = startDepositPeriod;
+        TENOR = tenor;
```

Comment:
Is there a better way to determine investment maturity? This seems like the tenor concept is a bit outside its area.